src/v/archival/ntp_archiver_service.cc

```diff
@@ -2785,24 +2795,51 @@ ntp_archiver::find_reupload_candidate(manifest_scanner_t scanner) {
           segment_collector_mode::collect_non_compacted);
         auto candidate = co_await collector.make_upload_candidate(
           _conf->upload_io_priority, _conf->segment_upload_timeout);
-        if (candidate.candidate.exposed_name().empty()) {
-            vlog(_rtclog.warn, "Failed to make upload candidate");
-            co_return std::make_pair(std::nullopt, std::nullopt);
-        }
-        if (
-          candidate.candidate.content_length != run->meta.size_bytes
-          || candidate.candidate.starting_offset != run->meta.base_offset
-          || candidate.candidate.final_offset != run->meta.committed_offset) {
-            vlog(
-              _rtclog.error,
-              "Failed to make upload candidate to match the run, candidate: "
-              "{}, "
-              "run: {}",
-              candidate.candidate,
-              run->meta);
-            co_return std::make_pair(std::nullopt, std::nullopt);
-        }
-        co_return std::make_pair(std::move(units), std::move(candidate));
+
+        using ret_t = std::pair<
+          std::optional<ssx::semaphore_units>,
+          std::optional<upload_candidate_with_locks>>;
+        co_return ss::visit(
+          candidate,
+          [](std::monostate) -> ret_t {
+              vassert(
+                false,
+                "unexpected default re-upload candidate creation result");
+          },
+          [this, &run, units = std::move(units)](
+            upload_candidate_with_locks& upload_candidate) mutable -> ret_t {
+              if (
+                upload_candidate.candidate.content_length
+                  != run->meta.size_bytes
+                || upload_candidate.candidate.starting_offset
+                     != run->meta.base_offset
+                || upload_candidate.candidate.final_offset
+                     != run->meta.committed_offset) {
+                  vlog(
+                    _rtclog.error,
+                    "Failed to make reupload candidate to match the run, "
+                    "candidate: "
+                    "{}, run: {}",
+                    upload_candidate.candidate,
+                    run->meta);
+                  return std::make_pair(std::nullopt, std::nullopt);
+              }
+
+              return std::make_pair(
+                std::move(units), std::move(upload_candidate));
+          },
+          [this](skip_offset_range& skip_offsets) -> ret_t {
+              vlog(
+                _rtclog.warn,
+                "Failed to make reupload candidate, skipping offsets: {}",
```

Comment:
This message is slightly confusing. It says skipping offsets, but I don't see anything that would lead to offsets being skipped. Bug? Needs rewording?

Reply:
this is possible if reupload races with truncation, the adjacent segment merger should just retry the operation